### PR TITLE
remove mandate of GCR creds for kaniko build

### PR DESCRIPTION
**Pull request by venkatk-25**
Currently Kaniko build requires configuring GCR credentials even if I am not using it,
 This change makes GCR creds optional and also makes adding other registry like ECR easy in future

### CODE REVIEW

**Comment by venkatk-25** [comment]
I have already made changes for skaffold fix in pkg/skaffold/schema/v1beta5/upgrade.go
Did I miss anything else. Should I move the logic to v1beta6 ?

**Comment by balopat** [comment]
> I have already made changes for skaffold fix in pkg/skaffold/schema/v1beta5/upgrade.go
> Did I miss anything else. Should I move the logic to v1beta6 ?

Oh, yeah, I missed that. It looks good :) 
And yes, as this introduces a config change, I merged `v1beta7` - so this change should land in that version, thus the upgrade will have to be in v1beta6 - please update it - it's a bit of an inconvenience - between the releases this always happens with the first config-change introducing PRs.

**Comment by venkatk-25** [comment]
I have made changes required and moves the fix to v1beta6

**Comment by dgageot** [comment]
@venkatk-25 Could you do one more rebase, please? 🥇

**Comment by venkatk-25** [comment]
@dgageot done

**Comment by balopat** [comment]
There is an integration test issue:

```
 REMOTE_INTEGRATION=true go test -v github.com/GoogleContainerTools/skaffold/integration -timeout 10m -run TestRun/kaniko 
```

will result in (after [this](#1800) fix):

```
Namespace: skaffoldd4q9l
--- FAIL: TestRun (0.19s)
    --- FAIL: TestRun/kaniko (0.08s)
        run_test.go:126: skaffold run: exit status 1, Generating tags...
             - gcr.io/k8s-skaffold/skaffold-example -> gcr.io/k8s-skaffold/skaffold-example:9eda3665a
            Tags generated in 14.450093ms
            Starting build...
            Creating kaniko secret [e2esecret]...
            time="2019-03-14T17:39:12-07:00" level=fatal msg="build failed: setting up pull secret: checking for existing kaniko secret: secrets \"e2esecret\" not found"
    --- FAIL: TestRun/kaniko_local (0.05s)
        run_test.go:126: skaffold run: exit status 1, Generating tags...
             - gcr.io/k8s-skaffold/skaffold-example -> gcr.io/k8s-skaffold/skaffold-example:9eda3665a
            Tags generated in 10.163924ms
            Starting build...
            Creating kaniko secret [e2esecret]...
            time="2019-03-14T17:39:12-07:00" level=fatal msg="build failed: setting up pull secret: checking for existing kaniko secret: secrets \"e2esecret\" not found"
    --- FAIL: TestRun/kaniko_local_-_sub_folder (0.06s)
        run_test.go:126: skaffold run: exit status 1, time="2019-03-14T17:39:12-07:00" level=warning msg="config version (skaffold/v1beta4) out of date: upgrading to latest (skaffold/v1beta7)"
            Generating tags...
             - gcr.io/k8s-skaffold/skaffold-example-sub -> gcr.io/k8s-skaffold/skaffold-example-sub:9eda3665a
            Tags generated in 20.163086ms
            Starting build...
            Creating kaniko secret [e2esecret]...
            time="2019-03-14T17:39:12-07:00" level=fatal msg="build failed: setting up pull secret: checking for existing kaniko secret: secrets \"e2esecret\" not found"
FAIL
```

File: pkg/skaffold/build/kaniko/sources/sources.go [hunk 1]
```diff
@@ -61,35 +61,56 @@ func podTemplate(cfg *latest.KanikoBuild, args []string) *v1.Pod {
...
-						SecretName: cfg.PullSecretName,
-					},
-				},
-			},
-			},
+			Volumes:       []v1.Volume{},
 		},
 	}
 
+	if cfg.GoogleCloudConfig == nil {
```
Comment (balopat) [inline pkg/skaffold/build/kaniko/sources/sources.go]:
```suggestion
	if cfg.GoogleCloudConfig != nil {
```

File: pkg/skaffold/build/kaniko/sources/sources.go [hunk 2]
```diff
@@ -61,35 +61,56 @@ func podTemplate(cfg *latest.KanikoBuild, args []string) *v1.Pod {
...
+		return addGoogleCloudConfig(pod, cfg.GoogleCloudConfig)
+	}
+
 	if cfg.DockerConfig == nil {
```
Comment (balopat) [inline pkg/skaffold/build/kaniko/sources/sources.go]:
```suggestion
	if cfg.DockerConfig != nil {
```
and we should cover this in a unit test too if possible

File: pkg/skaffold/build/kaniko/sources/sources.go (hunk 1 above)
Comment (balopat) [reply pkg/skaffold/build/kaniko/sources/sources.go]: and we should cover this in a unit test too if possible

**Comment by tejal29** [comment]
@venkatk-25 , in another PR #1700, @azaiter is proposing another way to specify ECR secret.
Maybe you both can collaborate on a config schema which supports other Container Registry which kaniko might support in future.
WDYT?
See https://github.com/GoogleContainerTools/skaffold/pull/1700#issuecomment-475068323 for more details.

**Comment by venkatk-25** [comment]
@tejal29 @balopat @azaiter I have created a new proposal https://github.com/GoogleContainerTools/skaffold/issues/1892 for the same. Can you go through and let me know your thoughts so I can take it forward and work on it further.

**Comment by tejal29** [comment]
@venkatk-25 if this is not the likely direction you are going to head (after #1892) , please close this PR.

**Comment by venkatk-25** [comment]
sure, closing this.

**Pull request closed without merging**